avsomr/avsomr.h

```diff
@@ -0,0 +1,182 @@
+#ifndef AVS_AVSOMR_H
+#define AVS_AVSOMR_H
+
+#include <memory>
+#include <QList>
+#include <QHash>
+#include <QSet>
+#include <QRect>
+#include <QImage>
+#include <QByteArray>
+
+namespace Ms {
+namespace Avs {
+
+class MsmrFile;
+class AvsOmr
+      {
+   public:
+      AvsOmr();
+      using ID = uint32_t;
+      using Num = uint16_t;
+      using Idx = uint16_t;
+
+      // Model
+      struct Barline {
+            ID id{0};
+            ID glyphID{0};
+            };
+
+      struct Staff {
+            ID id{0};
+            QList<ID> barlines;
+
+            struct {
+                  int32_t start{0};
+                  int32_t stop{0};
+                  ID clefID{0};
+                  ID keyID{0};
+                  ID timeID{0};
+                  } header;
+
+            bool isValid() const { return id > 0; }
+            };
+
+      struct MStack {
+            ID id{0};
+            Idx idx{0}; // resolved
+            int32_t left{0};
```

Comment:
Are things like `int32_t` really necessary? Why not just `int`?

Reply:
Well, int can be 16, 32 or 64 bit long, depending on platform

Reply:
> Are things like int32_t really necessary? Why not just int?  

@Howard-C I consider it a good style to explicitly indicate the type (int8_t, uint16_t, uint64_t, etc.). And right, just int different on different platforms. MU mobile app uses libmscore. And we encountered errors, though not with int, but with char. In some places, char is used. By default, the char can be signed or unsigned. And under Android it is unsigned, and -4 was written into it, as a result, the value became 252, instead of -4 :) (coordinates for drawing)

Reply:
Indeed, char needs to explicitly get qualified signed or insigned if it can get negative resp. can contain values larger than 127. A very common pitfall.
But do the corrent mobile apps really still use libmscore? I though they were written completly from scratch?
Anyway, we digress. using int32_t is OK, if it matters to have exactly 32bits, if not, is is just superfluous and plain int does everthing needed.

Reply:
> But do the corrent mobile apps really still use libmscore? I thought they had been rewritten completly from scratch?

Yes, the mobile application uses lib, but only it, all other systems, including rendering (implements QPaintEngine), view, audio system, synthesizer, etc. his own.

Reply:
> Anyway, we digress. using int32_t is OK, if it matters to have exactly 32 bits, if not, is is just superfluous (but harmless) and plain int does everthing needed. 
  
One of the reasons I prefer to specify the type explicitly is because I like the Golang. Golang was developed for the productive development of robust applications, taking into account design errors of other languages, especially C and C ++. In Go there is no "int", there are explicit types of int8, uint16, int32, etc. :) I trust the experience of architects Golang.

Reply:
I just don't quite understand why you'd want to make things more complex. Do you know some specific occasions where plain `int` isn't working? If not, then there's no reason to use `int32_t`, whether or not some other softwares/frameworks/companies/whatever use it. This is MuseScore, I think we should follow the conventions already established here (unless an authorized change is made).

Reply:
> I just don't quite understand why you'd want to make things more complex.  
  
On the contrary, I want to make the code simpler and more consistent.
The use of fixed types makes the developer think about what needs to be used in a particular case, for example:
* If this is an object field that has many instances, then you should strive to use the smallest suitable type to save memory.
* If it is byte size or time, then you need to use the maximum available type so as not to go beyond the boundaries.
* Never use a char ever to store a numeric value.
* etc.

If you get used to just int, then everywhere this int and write, without thinking about what you really need.
A simple practice of using fixed types, we can avoid several classes of errors and add understanding. Isn't it better to use it?

Reply:
Maybe.

Reply:
Using the smallests type that still fits the need indeed does save on memory, but looses on performance. In general an plain 'int', at whatever the implementation defines its size to be, is the one it performs best on, it is (supposit to be) the systems 'native' size

Reply:
Yes, there is a difference in performance, but this difference matters in those places where a huge number of mathematical operations take place, this is not a very large limited number of places, is it not so?  That is, does anyone have a benchmark showing the difference in performance?
A class of errors associated with types is more common and more dangerous, isn’t it?

Reply:
What errors specifically speaking? This is `int`, not vectors which may have go-beyond-boundary errors, or `char` which may has other problems.

Reply:
I'm talking in general here, not picking on that special case. And no, I don't have benchmarks.
But I know Werner  was removing (or at least frowen upon) some `enum class XXX : char` (added by me a long time ago, to indeed save space/memory) for such performance reasons

Reply:
> What errors specifically speaking? This is `int`, not vectors which may have go-beyond-boundary errors, or `char` which may has other problems.

Inside structs better be explicit about sizes, for reasone of padding and memory alignments

Reply:
But it isn't some kind of "unknown" value, `int` is defined as 32-bit in some OS's and 16-bit or 64-bit in others, they're different but not undetermined.

Reply:
> This is int, not vectors which may have go-beyond-boundary errors  

As I wrote above, if you get used to using fixed types, you always think which type to use, and why. If you get used to using just int, then there is a high probability that this will lead to an error.

> Inside structs better be explicit about sizes, for reasone of padding and memory alignments
another good reason to always think about the type used :)

Reply:
Only thing the standards quarantee is char <= short <= int <= long <= long long && char >= 8 bit && sizeof (char) == 1. In paxis though 16-bit ints have gone out of fashion at least a decade ago and 64-bit int are not used, but long or long long instead.

Reply:
> I'm talking in general here, not picking on that special case.  

using int instead of fixed type for performance - maybe it's antipattern [Premature optimization](https://en.wikipedia.org/wiki/Program_optimization#When_to_optimize)?

Reply:
> then there is a high probability that this will lead to an error

Again, what error, exactly? If you can name at least one, I will let go of this matter.

Reply:
I will name the typical errors that I encountered:
* type overflow - this can happen when more than a value is stored in the type (size in bytes, time in nanoseconds, etc.). Or when calculations occur, and intermediate values can be large.
* memory overflow - due to the use of unreasonably large types or due to unbalanced memory in structures.
* design errors - it is not indicated explicitly whether the value can be negative or not, that is, they often write just `int`, not `unsigned int`, simply because they are used to

Reply:
So you're suggesting we should specify type size everywhere we use `int`? Since there may be "so many errors"? You can put this in the proposal page too, I encourage you to do so.

Reply:
I believe this discission does not belong here.

Reply:
> So you're suggesting we should specify type size everywhere we use int? Since there may be "so many errors"?  

At a minimum, I suggesting you not to prohibit this.
For example, for me, clarity are always better than ambiguity, for me it is an axiom :)

I will make a proposal to change this aspect

Reply:
> for me, clarity are always better than ambiguity

That's true whatsoever. I'll stay neutral on this matter.

Reply:
It certainly is not forbidden. In a `for (uint32_t i =0; ..., ; i++)` loop though ...

Reply:
In loop, most often you need to use `size_t` :)

Reply:
if looping through something that is an arrack of sorts, yes